### ISSUE TITLE
[LAB] Week 2_Problem 1_Hajin KIM

### ISSUE THREAD

**Issue by khj021101** [vlab-kaist/NN101_23S#88]
import torch
from random import random 
from typing import Callable

##                                                            ##

# NOTE : Feel free to use torch.optim and tensor.

def training(x_train : list[float], y_train : list[float]) -> list[float]: # DO NOT MODIFY FUNCTION NAME
    # Data normalization code (prevents overflow when calculating MSE, prevents underfitting)
    # Note that you need to convert [w, b] to the original scale before returning value
    # w = w * (y_max - y_min)
    # b = b * (y_max - y_min) + y_min
    y_min = min(y_train)
    y_max = max(y_train)
    normalize = lambda y : (y - y_min)/(y_max - y_min) #y값은 0~1 사이로 압축한 뒤 계산.
    ### IMPLEMENT FROM HERE
    w_init = (normalize(y_train[1])-normalize(y_train[0]))/(x_train[1]-x_train[0]) #초기값
    b_init = normalize(y_train[0]) - w_init*x_train[0] #초기값
    elnum = len(x_train) #원소개수
    norm_y = y_train #normalize 한 ytrain
    for i in range(elnum):
        norm_y[i] = normalize(y_train[i])
    
    def loss(tan : float, y0 : float) -> float:
        res = 0
        for i in range(elnum):
            res += (tan*x_train[i]+y0-norm_y[i])**2
        err = (1/(2*elnum))*res
        return err
    def wdiff(tan: float, y0 : float) -> float:
        res = 0 
        for i in range(elnum):
            res += ((tan*x_train[i])+y0-norm_y[i])*x_train[i]/elnum
        return res
    def bdiff(tan: float, y0: float) -> float:
        res = 0
        for i in range(elnum):
            res += ((tan*x_train[i])+y0-norm_y[i])/elnum
        return res
    stepsize = 0.05
    w = w_init
    b = b_init
    for i in range(100):
        w = w - stepsize*wdiff(w, b)
        b = b - stepsize*bdiff(w, b)
        #print('loss = ' + str(loss(w,b)))#loss 확인
        #print(w, b)
    return[w,b]

def predict(x_train : list[float], y_train : list[float], x_test : list[float]) -> list[float]: # DO NOT MODIFY FUNCTION NAME
    [w,b] = training(x_train, y_train)
    y_test = x_test
    for i in range(len(x_test)):
        y_test[i] = x_test[i]*w + b
    return y_test

if __name__ == '__main__':
    x_train = [0.0, 1.0, 2.0, 3.0, 4.0]
    y_train = [3.0, 1.0, 6.0, 7.0, 10.0] # Note that not all test cases give clear line.
    x_test = [5.0, 10.0, 8.0]
    
    w, b = training(x_train, y_train)
    y_test = predict(x_train, y_train, x_test)

    print(w, b)
    print(y_test)

**Comment by Dongyeongkim** [comment]
Rewrite another issue. the form is broken. will be resolved.
